### PR TITLE
Utility/DatabaseFilter

### PR DESCRIPTION
**Pull request by oliveralka**
Tool to create a reduced database from a standard FASTA database.

A FASTA database serves as input and will be filtered depending on a give reverence vector.
Given entries are pasted to the new reduced database.
Filter whitelist: Input whitelist as reverence; Output database with entries on whitelist.
Filter blacklist: Input blacklist as reverence; Output database without entries on blacklist.

### CODE REVIEW

File: src/utils/DatabaseFilter.cpp [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+//-------------------------------------------------------------
+//Doxygen docu
+//-------------------------------------------------------------
+
+/**
+    @page UTILS_DatabaseFilter DatabaseFilter
+
+    @brief Tool to create a reduced database from a standard FASTA database.
+
+    A FASTA database serves as input and will be filtered depending on a give reverence vector.
```
Comment (timosachsenberg) [inline src/utils/DatabaseFilter.cpp]: maybe rephrase: "depending on a give reverence vector.Given entries are pasted to the new reduced database."
to something like:

The DatabaseFilter tool filters a protein database in fasta format according to one or multiple filtering criteria.

The resulting database is written as output, Depending on the reporting method (method="whitelist" or "blacklist") only entries are retained that passed all filters ("whitelist) or failed at least one filter ("blacklist").

Implemented filter criteria:
- accession: Filter database according to the set of protein accessions contained in an identification file

File: src/utils/DatabaseFilter.cpp [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+  {
+
+  }
+
+protected:
+  void registerOptionsAndFlags_()
+  {
+    registerInputFile_("in", "<file>", "","Input FASTA file, containing a database.");
+    setValidFormats_("in", ListUtils::create<String>("fasta"));
+    registerInputFile_("accession", "<file>", "", "Input IdXML file, containing the identfied peptides.", true);
```
Comment (timosachsenberg) [inline src/utils/DatabaseFilter.cpp]: add support for mzIdentML (alternative identification file format)

**Comment by timosachsenberg** [comment]
Nice addition. Only some minor comments from my side.

File: src/utils/DatabaseFilter.cpp (hunk 1 above)
Comment (hroest) [inline src/utils/DatabaseFilter.cpp]: did you mean "reference" instead of "reverence" ?

File: src/utils/DatabaseFilter.cpp [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+      }
+    }
+
+    LOG_INFO << "Protein accessions: " << id_accessions.size() << endl;
+
+    for (Size i = 0; i != db.size() ; ++i)
+    {
+      const String& fasta_accession = db[i].identifier;
+      const bool found = id_accessions.find(fasta_accession) != id_accessions.end();
+      if ( (found && whitelist) || (!found && !whitelist) )
```
Comment (hroest) [inline src/utils/DatabaseFilter.cpp]: maybe add a comment here: "either found in the whitelist or not found in the blacklist"

File: src/utils/DatabaseFilter.cpp [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  }
+
+protected:
+  void registerOptionsAndFlags_()
+  {
+    registerInputFile_("in", "<file>", "","Input FASTA file, containing a database.");
+    setValidFormats_("in", ListUtils::create<String>("fasta"));
+    registerInputFile_("accession", "<file>", "", "Input IdXML file, containing the identfied peptides.", true);
+    setValidFormats_("accession", ListUtils::create<String>("idXML"));
+    registerStringOption_("method", "<type>", "whitelist", "Switch between white/blacklisting", false);
```
Comment (hroest) [inline src/utils/DatabaseFilter.cpp]: please explain in description what this does. Often people will only look at the parameter description on CLI or in TOPPAS and therefore have trouble to understand what this parameter will do

File: src/utils/DatabaseFilter.cpp (hunk 1 above)
Comment (hroest) [inline src/utils/DatabaseFilter.cpp]: Only to be clear, the reference list is delivered as idXML ? maybe this should be clearly written somewhere

**Comment by hroest** [comment]
only minor comments, looks good to me. please merge when all are adressed

**Comment by oliveralka** [comment]
Edit description:
```
The DatabaseFilter tool filters a protein database in fasta format according to one or multiple filtering criteria.
The resulting database is written as output. Depending on the reporting method (method="whitelist" or "blacklist") only entries are retained that passed all filters ("whitelist) or failed at least one filter ("blacklist").
Implemented filter criteria:
        accession: Filter database according to the set of protein accessions contained in an identification file (idXML, mzIDentML)
```
Add support for mzIdentML:
```
if(ids_type == FileTypes::IDXML)
     {
       IdXMLFile().load(ids, protein_identifications, peptide_identifications);
     }
     else if(ids_type == FileTypes::MZIDENTML)
     {
       MzIdentMLFile().load(ids, protein_identifications, peptide_identifications);
     }
     else
     {
       writeLog_("Error: Unknown input file type given. Aborting!");
       printUsage_();
       return ILLEGAL_PARAMETERS;
     }
```
add comment: 
`if ( (found && whitelist) || (!found && !whitelist) ) //either found in the whitelist or not found in the blacklist
`

File: src/utils/DatabaseFilter.cpp [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    bool whitelist = (method == "whitelist");
+    String out(getStringOption_("out"));
+
+    //-------------------------------------------------------------
+    // reading input
+    //-------------------------------------------------------------
+
+    vector<FASTAFile::FASTAEntry> db;
+    FASTAFile ().load(in, db);
+
```
Comment (timosachsenberg) [inline src/utils/DatabaseFilter.cpp]: check if no filter has been specified. if so write out the original db

File: src/utils/DatabaseFilter.cpp [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    FileHandler fh;
```
Comment (timosachsenberg) [inline src/utils/DatabaseFilter.cpp]: put into if (!ids.empty()) block

File: src/utils/DatabaseFilter.cpp [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+      return ILLEGAL_PARAMETERS;
+    }
+
+    LOG_INFO << "Identifications: " << ids.size() << endl;
+
+    //-------------------------------------------------------------
+    // calculations
+    //-------------------------------------------------------------
+
+    vector<FASTAFile::FASTAEntry> db_new;
```
Comment (timosachsenberg) [inline src/utils/DatabaseFilter.cpp]: move above if(!ids.empty()) block

**Comment by oliveralka** [comment]
added check and filterByProteinAccessions function

File: src/utils/DatabaseFilter.cpp [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+  }
+
+protected:
+  void registerOptionsAndFlags_()
+  {
+    registerInputFile_("in", "<file>", "","Input FASTA file, containing a database.");
+    setValidFormats_("in", ListUtils::create<String>("fasta"));
+    registerInputFile_("accession", "<file>", "", "Input IdXML file, containing the identfied peptides.", true);
+    setValidFormats_("accession", ListUtils::create<String>("idXML,mzid"));
+    registerStringOption_("method", "<type>", "whitelist", "Switch between white/blacklisting", false);
```
Comment (hendrikweisser) [inline src/utils/DatabaseFilter.cpp]: For consistency with other tools and increased clarity:
`registerStringOption_("method", "<choice>", "whitelist", "Switch between white-/blacklisting of protein accessions", false);`

File: src/utils/DatabaseFilter.cpp [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+// We do not want this class to show up in the docu:
+/// @cond TOPPCLASSES
+
+class TOPPDatabaseFilter :
+  public TOPPBase
+{
+public:
+  TOPPDatabaseFilter() :
+    TOPPBase("DatabaseFilter", "The DatabaseFilter tool filters a protein database in fasta format according to one or multiple filtering criteria", false)
```
Comment (hendrikweisser) [inline src/utils/DatabaseFilter.cpp]: There's currently only one criterion, so this is a bit misleading/vague. Also, no need to include the tool name in the description. Maybe just:
"Filters a protein database (FASTA format) based on identified proteins"

File: src/utils/DatabaseFilter.cpp [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+    String method(getStringOption_("method"));
+    bool whitelist = (method == "whitelist");
+    String out(getStringOption_("out"));
+
+    //-------------------------------------------------------------
+    // reading input
+    //-------------------------------------------------------------
+
+    vector<FASTAFile::FASTAEntry> db;
+    FASTAFile ().load(in, db);
```
Comment (hendrikweisser) [inline src/utils/DatabaseFilter.cpp]: No space before "()".

File: src/utils/DatabaseFilter.cpp [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+    TOPPBase("DatabaseFilter", "The DatabaseFilter tool filters a protein database in fasta format according to one or multiple filtering criteria", false)
+  {
+  }
+
+protected:
+  void registerOptionsAndFlags_()
+  {
+    registerInputFile_("in", "<file>", "","Input FASTA file, containing a database.");
+    setValidFormats_("in", ListUtils::create<String>("fasta"));
+    registerInputFile_("accession", "<file>", "", "Input IdXML file, containing the identfied peptides.", true);
```
Comment (hendrikweisser) [inline src/utils/DatabaseFilter.cpp]: In other tools, we use "id" as the parameter name for a secondary input file containing ID results (e.g. IDMapper, FeatureFinderIdentification). I would prefer:
`registerInputFile_("id", "<file>", "", "Input file containing identified peptides and proteins.");`
(Then of course you'd have to replace "accession" everywhere.)

File: src/utils/DatabaseFilter.cpp [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+  {
+    set<String> id_accessions;
+    for (Size i = 0; i != peptide_identifications.size(); ++i)
+    {
+      const PeptideIdentification& id = peptide_identifications[i];
+      const vector<PeptideHit>& hits = id.getHits();
+      for (Size k = 0; k != hits.size(); ++k)
+      {
+        const vector<PeptideEvidence>& evidences = hits[k].getPeptideEvidences();
+        for (Size m = 0; m != hits.size(); ++m)
```
Comment (hendrikweisser) [inline src/utils/DatabaseFilter.cpp]: `evidences.size()`, not `hits.size()`!

File: src/utils/DatabaseFilter.cpp [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+      }
+    }
+
+    LOG_INFO << "Protein accessions: " << id_accessions.size() << endl;
+
+    for (Size i = 0; i != db.size() ; ++i)
+    {
+      const String& fasta_accession = db[i].identifier;
+      const bool found = id_accessions.find(fasta_accession) != id_accessions.end();
+      if ( (found && whitelist) || (!found && !whitelist) ) //either found in the whitelist or not found in the blacklist
```
Comment (hendrikweisser) [inline src/utils/DatabaseFilter.cpp]: No spaces between two opening/closing brackets.

File: src/utils/DatabaseFilter.cpp (hunk 11 above)
Comment (timosachsenberg) [inline src/utils/DatabaseFilter.cpp]: We also had this once but thought this might be slightly confusing because the is the "id" part in a fasta entry.  No strong opinion about this so if you consider this important I think Oliver can change it back.

File: src/utils/DatabaseFilter.cpp (hunk 12 above)
Comment (timosachsenberg) [inline src/utils/DatabaseFilter.cpp]: good catch

**Comment by oliveralka** [comment]
Changed accession to id and other minor changes.

File: src/utils/DatabaseFilter.cpp [hunk 14]
```diff
@@ -56,7 +56,7 @@ using namespace std;
 
     Implemented filter criteria:
 
-        accession: Filter database according to the set of protein accessions contained in an identification file (idXML, mzIdentML)
+        ID: Filter database according to the set of proteinIDs contained in an identification file (idXML, mzIdentML)
```
Comment (timosachsenberg) [inline src/utils/DatabaseFilter.cpp]: I would keep "protein accession" as we don't filter according to protein_identifications but the protein accessions contained in the peptide_identifications

File: src/utils/DatabaseFilter.cpp [hunk 15]
```diff
@@ -81,15 +81,15 @@ class TOPPDatabaseFilter :
   {
     registerInputFile_("in", "<file>", "","Input FASTA file, containing a database.");
     setValidFormats_("in", ListUtils::create<String>("fasta"));
-    registerInputFile_("accession", "<file>", "", "Input IdXML file, containing the identfied peptides.", true);
-    setValidFormats_("accession", ListUtils::create<String>("idXML,mzid"));
-    registerStringOption_("method", "<type>", "whitelist", "Switch between white/blacklisting", false);
+    registerInputFile_("id", "<file>", "", "Input file containing identified peptides and proteins.");
+    setValidFormats_("id", ListUtils::create<String>("idXML,mzid"));
+    registerStringOption_("method", "<choice>", "whitelist", "Switch between white-/blacklisting of protein IDs", false);
```
Comment (timosachsenberg) [inline src/utils/DatabaseFilter.cpp]: remove "of protein IDs" as we will use black / whitelisting for different filters too

File: src/utils/DatabaseFilter.cpp [hunk 16]
```diff
@@ -81,15 +81,15 @@ class TOPPDatabaseFilter :
...
     setValidStrings_("method", ListUtils::create<String>("whitelist,blacklist"));
     registerOutputFile_("out", "<file>", "", "Output FASTA file where the reduced database will be written to.");
     setValidFormats_("out", ListUtils::create<String>("fasta"));
   }
 
-  void filterByProteinAccessions_(const vector<FASTAFile::FASTAEntry>& db, const vector<PeptideIdentification>& peptide_identifications, bool whitelist, vector<FASTAFile::FASTAEntry>& db_new)
+  void filterByProteinIDs_(const vector<FASTAFile::FASTAEntry>& db, const vector<PeptideIdentification>& peptide_identifications, bool whitelist, vector<FASTAFile::FASTAEntry>& db_new)
```
Comment (timosachsenberg) [inline src/utils/DatabaseFilter.cpp]: if you want to emphasize that IDs are used then ist should read filterByPeptideIDs_

**Comment by timosachsenberg** [comment]
Sorry my comments are a bit nitpicking.
I think the confusion arises from the fact that ID has several meanings:
- the accession in a fasta file
- Identification in general (could be protein ident. or peptide ident.)

File: src/utils/DatabaseFilter.cpp (hunk 8 above)
Comment (timosachsenberg) [inline src/utils/DatabaseFilter.cpp]: Agree but remove " of protein accessions". The white / blacklisting is independent of the filter criterium. There more to come and we don't want to change parameter documentation all the time.

**Comment by hendrikweisser** [comment]
@timosachsenberg: I think you and I are giving conflicting advice about how general or specific descriptions of the tool, of parameters etc. should be. I get the impression that you have plans to extend the functionality of this tool, but how concrete are these and what's the timeline?
I think a tool that fulfils a very specific function, like this one currently does, should not have overly general descriptions because the functionality may be extended in the future.

**Comment by hroest** [comment]
yes I think we should not nitpick too much here and merge and see how it goes ...

**Comment by timosachsenberg** [comment]
No problem Hendrik. We can keep the line you proposed. The other comments are in my view more important because we have a ProteinIdentification class that can easily be confused with how Oliver now used the term ProteinID for protein accessions.

**Comment by oliveralka** [comment]
"id" is now used for the registerInputFile.
For the other variables and comments "accession" is used 
- I think this is a good compromise.

**Comment by hendrikweisser** [comment]
@oliveralka: Thanks for being patient and making all those changes.

I now realize that my comment "(Then of course you'd have to replace "accession" everywhere.)" added to the confusion - I only meant the variable name "accession".

**Comment by timosachsenberg** [comment]
Thanks @oliveralka  and welcome on board :)
